hs_core/views/resource_folder_hierarchy.py

```diff
@@ -1,456 +1,697 @@
+    Thus it is much more limiting than a general purpose REST responder.
+    """
+    pk = request.POST.get('res_id', pk)
+    if pk is None:
+        return HttpResponse('Bad request - resource id is not included',
+                            status=status.HTTP_400_BAD_REQUEST)
+    pk = str(pk).strip()
+    try:
+        resource, _, user = authorize(request, pk,
+                                      needed_permission=ACTION_TO_AUTHORIZE.EDIT_RESOURCE)
+    except NotFound:
+        return HttpResponse('Bad request - resource not found', status=status.HTTP_400_BAD_REQUEST)
+    except PermissionDenied:
+        return HttpResponse('Permission denied', status=status.HTTP_401_UNAUTHORIZED)
+
+    tgt_path = resolve_request(request).get('target_path', None)
+    src_paths = resolve_request(request).get('source_paths', None)
+    if src_paths is None or tgt_path is None:
+        return HttpResponse('Bad request - src_paths or tgt_path is not included',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    tgt_path = str(tgt_path).strip()
+    if not tgt_path:
+        return HttpResponse('Target directory not specified',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    # protect against common hacking attacks
+    if not tgt_path.startswith('data/contents/'):
+        return HttpResponse('Target directory path must start with data/contents/',
+                            status=status.HTTP_400_BAD_REQUEST)
+    if tgt_path.find('/../') >= 0 or tgt_path.endswith('/..'):
+        return HttpResponse('Bad request - tgt_path cannot contain /../',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    istorage = resource.get_irods_storage()
+    tgt_abspath = os.path.join(resource.root_path, tgt_path)
+    tgt_abspath = tgt_abspath.rstrip('/')
+
+    if not irods_path_is_directory(istorage, tgt_abspath):
+        return HttpResponse('Target of move is not an existing folder',
+                            status=status.HTTP_400_BAD_REQUEST)
+
+    src_paths = json.loads(src_paths)
+
+    # protect against common hacking attacks
+    for src_path in src_paths:
+        src_path = str(src_path).strip()
+
+        if not src_path.startswith('data/contents/'):
+            return HttpResponse('Paths to be moved must start with data/contents/',
+                                status=status.HTTP_400_BAD_REQUEST)
+
+        if src_path.find('/../') >= 0 or src_path.endswith('/..'):
+            return HttpResponse('Paths to be moved cannot contain /../',
+                                status=status.HTTP_400_BAD_REQUEST)
+
+    # strip trailing slashes (if any)
+    tgt_path = tgt_path.rstrip('/')
+    for i in range(len(src_paths)):
+        src_paths[i] = src_paths[i].rstrip('/')
+
+    for s in src_paths:
+        abs_path = os.path.join(resource.root_path, s)
+
+        # protect against stale data botches: source files should exist
```

Comment:
Great!! Looks like we have to make similar changes many other places where we are allowing the user to manipulate files. I will see if I need to make similar checks in my file type work.